### PR TITLE
dce_gettime now returns the node time

### PR DESCRIPTION
**Pull request by teto**
ns3 introduced Node::GetLocalTime to return a node specific time. This
leverages this feature.

I've added some documentation more or less in a  doxygen style. Let me know if that's fine.
This should work as is but if you prefer, you may want to leave this PR pending until I get a full patch with ntp examples etc...

### CODE REVIEW

File: model/utils.cc [hunk 1]
```diff
@@ -153,6 +159,31 @@ struct timespec UtilsTimeToTimespec (Time time)
   tv.tv_nsec = n % 1000000000;
   return tv;
 }
+
+std::string
+UtilsGenerateIfNameFromIndex(uint32_t i) {
```
Comment (thehajime) [inline model/utils.cc]: please describe this addition to the commit message.

**Comment by thehajime** [comment]
please also add a description to RELEASE_NOTES (in new user-visible features) so that others can recognize this.

overall it looks fine.
I would suggest to add a _minimum_ test for this new feature before adding full patch including ntpd support.

**Comment by teto** [comment]
I used the UtilsGenerateIfNameFromIndex when I was fixing netlink (I am trying to backport little by little hundreds of modifications). I will try to send PRs about netlink before fixing this one.

**Comment by teto** [comment]
When updating this PR, I wonder if you would agree to merge stub implementations of the following syscalls:
====
DCE (clock_settime)
DCE_EXPLICIT (clock_getres, int, clockid_t, struct timespec *)

// TIMEX.H
DCE(adjtimex) // Specific to linux I believe so maybe not use it ?
NATIVE_WITH_ALIAS2(adjtime, ntp_adjtime) 
===
The goal is then to implement these stub functions (and I have a preliminary Gsoc implementation) but having the prototype already declared would help me since I have to go back & forth between different/intertwined ns3 and DCE branches & PRs.

Only difference for the user is that in one case DCE does not find the syscall, in the other it will throw a FATAL_ERROR.

**Comment by thehajime** [comment]
> When updating this PR, I wonder if you would agree to merge stub implementations of the following syscalls:

I think it's fine to merge it. do you want to update this PR to include this or will have a separate PR ?

I will merge this PR if I got a response from you.

**Comment by teto** [comment]
I am trying to go as far as I can before ns3.25 get released. If I can't finish for dce 1.8 I will push stub implementations.

I also have a question about some code you added in  commit c55a7c73413b048be853ff65ded11d098c20bfed:
```
GlobalValue g_timeBase = GlobalValue ("SimulationTimeBase",
                                      "The timebase for this simulation",
                                      // January 1st, 2010, 00:00:00
                                      UintegerValue (1262304000),
                                      MakeUintegerChecker<uint32_t> ());
```
Why did you implement this ? Can't it be achieved via starting the simulation later Simulator::Start( Seconds(1262304000)). It creates inconsistencies between ns3 and dce for instance when generating pcaps, the times do not match which is a problem in what I am currently doing. My take would be either to remove it (depending on backward compatibility concern) or to move it to ns3.

**Comment by thehajime** [comment]
> I am trying to go as far as I can before ns3.25 get released. If I can't finish for dce 1.8 I will push stub implementations.

okay.

> I also have a question about some code you added in commit c55a7c7:
> 
> GlobalValue g_timeBase = GlobalValue ("SimulationTimeBase",
>                                       "The timebase for this simulation",
>                                       // January 1st, 2010, 00:00:00
>                                       UintegerValue (1262304000),
>                                       MakeUintegerChecker<uint32_t> ());
> Why did you implement this ? Can't it be achieved via starting the simulation later Simulator::Start( Seconds(1262304000)). It creates inconsistencies between ns3 and dce for instance when generating pcaps, the times do not match which is a problem in what I am currently doing. My take would be either to remove it (depending on backward compatibility concern) or to move it to ns3.

this was originally from the most earliest version of dce (see https://github.com/direct-code-execution/ns-3-dce/blob/b3533d68f0239ae14a90fa00fdf9fcb7ae3d7914/model/utils.cc#L136), which was not coded by myself.

but I can guess that there are programs which cannot run properly IF time function (e.g., ```gettimeofday()```) returns 00:00:00 Jan 1 1970. I at least know _the_ ```ping``` command in iputils can't send/recv packets when the ```tv_sec==0``` obtained from ```gettimeofday```.

https://github.com/iputils/iputils/blob/master/ping_common.c#L322

**Comment by teto** [comment]
This needs a lot more work than I thought. I close this and will open other PRs when it's more mature.

**Comment by thehajime** [comment]
ack.

**Pull request closed without merging**